### PR TITLE
#159053991 Allow Users to Favourite and  Unfavourite Article

### PR DESCRIPTION
**Pull request by munyili-samuel**
#### What does this PR do
- Allows users to favourite and unfavourite an article

#### Description of tasks done
- added DB models for favouriting article
- Added urls endpoint for favouriting article
- added views to handle favouriting and un favouriting
- made migrations for database model changes
- added unit tests for this functionality

#### How can this PR be manually tested
Prerequisite: Postman
- use git pull to get the source code from ft-favourite-article-159053991 branch
- use `source .env` to export environmental variables and activate virtual env
- run `python manage.py migrate` to apply recent migrations
- run python manage.py runserver` to start  the project
- if not already done, create a new user account and login
- if not already done, create an article and copy the `slug`
- create a `POST` request to the following endpoint `<str:slug>/favorite/` replacing the `<slug>` with the slug you copied above
- you should get the article back. If you try again you should get a message `You have already marked this article as a favourite`
- create a `DELETE` request to the following endpoint `<str:slug>/favorite/` replacing the `<slug>` with the slug you copied above
- you should get the article back. If you try again you should get a message `You have not marked this article as a favourite`
- if you use a slug that is not valid, you should get a message `The article was not found` on both requests

#### Related PT stories
- [#159053991](https://www.pivotaltracker.com/story/show/159053991)

### CODE REVIEW

**Review by hoslack — APPROVED** [review]

**Review by RuthNjeri — CHANGES_REQUESTED** [review]
Hi Sam,
When I favourite an article using a POST request
```
http://127.0.0.1:8000/api/v1/articles/how-to-train-the-dragon-in-a-ditch-7895168a9cbd41b8ac1822cb729766c0/favorite/
```
 I am getting this as the response 
``` 
{
    "article": {}
}
```
and when unfavourite an article, I get this back
```
{
    "message": {}
}
```
Shouldn't I be getting the article back?

**Comment by RuthNjeri** [comment]
When I `GET an article` after favoriting it, I get this response
```
    "favorited": [
        2
    ]
```
Is there a reason you have chosen to return the user id of a person who has favorited an article as opposed to their username?

**Comment by munyili-samuel** [comment]
I will fix this in the next 30min.

**Review by davenjeru — APPROVED** [review]
The PR has been tested and is okay. It has been stated that the end user does not need to know who has favourited an article, they just need to see their favourited articles.

**Comment by davenjeru** [comment]
It is not a big deal, but who else feels weary about the use of American English on the API endpoints 😅😅😅, Like **favorite** instead of **favourite**. 🤷‍♂️

File: authors/apps/articles/views.py [hunk 1]
```diff
@@ -112,3 +108,82 @@ def update(self, request, slug):
...
+    """
+
+    permission_classes = (IsAuthenticated,)
+    queryset = Article.objects.all()
+    serializer_class = ArticleSerializer
+
+    def post(self, request, slug):
+        """
+        Helps user favourite article
+        If already, return message saying that the user has already favourated
```
Comment (lennykioko) [inline authors/apps/articles/views.py]: Typo: you meant favourited instead of favourated

**Review by RuthNjeri — APPROVED** [review]

**Review by dmigwi — CHANGES_REQUESTED** [review]
Please fix all the pep8 violations on your PR.

**Comment by dmigwi** [comment]
How did you end up making 87 commits?

**Pull request merged**